### PR TITLE
dummy login

### PR DESCRIPTION
**Pull request by mchalski**
MEN-795

@bboozzoo @kjaskiewiczz not really ready for merge (licenses fail, cosmetic changes) but you can start reviewing

### CODE REVIEW

File: api_useradm.go [hunk 1]
```diff
@@ -53,14 +67,39 @@ func (i *UserAdmHandlers) GetApp() (rest.App, error) {
...
+
+	//parse auth header
+	//TODO: we'll allow no auth in the 'first login' case, for now - just return 401
+	email, pass, ok := r.BasicAuth()
+	if !ok {
+		rest_utils.RestErrWithLog(w, r, l, ErrAuthHeader, http.StatusUnauthorized)
+		return
+	}
+
+	useradm, err := u.createUserAdm(config.Config, l)
```
Comment (bboozzoo) [inline api_useradm.go]: I'm wondering if it's possible to get rid of this pattern. The config does not change between requests. Looks like logger is the only thing left. Perhaps we could pack this into a `context.Context`? Then some helper:

```
func LoggerFromContext(ctx context.Context) {
    l, ok := ctx.Value(userKey).(*log.Logger)
    if !ok {
        return log.New(log.Ctx{})
    } 
    return l
}
```
Then when calling NewUserAdmApiHandlers() pass a closure that captures config. So the only thing left here would be (error goes away, as config is not used here anymore?):

```
useradm := u.UserAdm()
...
```

File: useradm.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+type UserAdm struct {
+	jwtHandler JWTHandler
+}
+
+func NewUserAdm(jwtHandler JWTHandler) *UserAdm {
+	return &UserAdm{jwtHandler: jwtHandler}
+}
+
+func GetUserAdm(c config.Reader, l *log.Logger) (UserAdmApp, error) {
```
Comment (bboozzoo) [inline useradm.go]: `l` is not used?

File: jwt.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+// JWTHandlerRS256 is an RS256-specific JWTHandler
+type JWTHandlerRS256 struct {
+	issuer       string
+	expiresInSec int64
+	privKey      *rsa.PrivateKey
+	log          *log.Logger
+}
+
+func NewJWTHandlerRS256(privKeyPath, issuer string, expiresInSec int64) (*JWTHandlerRS256, error) {
+	priv, err := getRSAPrivKey(privKeyPath)
```
Comment (bboozzoo) [inline jwt.go]: Change to `NewJWTHandlerRS256(privKey *rsa.privateKey, issuer string, expiresInSec int64) *JWTHandler`

Shouldn't this accept a logger as well?

File: api_useradm.go [hunk 4]
```diff
@@ -25,15 +30,24 @@ const (
 	uriUsersInitial = "/api/0.1.0/users/initial"
 )
 
-type UserAdmHandlers struct {
+var (
+	ErrAuthHeader = errors.New("invalid or missing auth header")
+)
+
+type UserAdmFactory func(c config.Reader, l *log.Logger) (UserAdmApp, error)
```
Comment (bboozzoo) [inline api_useradm.go]: Change to:
`type UserAdmFactory func (l *log.Logger) UserAdmApp`

File: api_useradm.go (hunk 1 above)
Comment (bboozzoo) [inline api_useradm.go]: Once `NewJWTHandler..` stops returning errors and `config.Config` is removed from `UserAdmFactory` this piece becomes `useradm := u.createUserAdm(l)`

File: server.go [hunk 5]
```diff
@@ -40,14 +40,14 @@ func RunServer(c config.Reader) error {
 
 	l := log.New(log.Ctx{})
 
-	invapi := NewUserAdmApiHandlers()
+	useradmapi := NewUserAdmApiHandlers(GetUserAdm)
```
Comment (bboozzoo) [inline server.go]: change to 
```
// load private key
privKey, err := getRSAPrivKey(c.GetString(SettingPrivKeyPath))
// error handling
useradmapi := NewUserAdmApiHandlers(func (l *log.Logger) UserAdmApp {
	jwtHandler := NewJWTHandlerRS256(
		privKey,
		c.GetString(SettingJWTIssuer),
		int64(c.GetInt(SettingJWTExpirationTimeout)),
		l) // pass a logger to jwt handler as well, see comment about NewJWTHandler..

	return NewUserAdm(jwtHandler)
})
```

Comment (bboozzoo) [inline server.go]: BTW.  `GetUserAdm` is no longer needed

File: jwt.go [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+	jwtClaimIssuer     = "iss"
+	jwtClaimSubject    = "sub"
+	jwtClaimExpiration = "exp"
+	jwtClaimTokenId    = "jti"
+)
+
+// JWTHandler jwt generator/verifier
+type JWTHandler interface {
+	GenerateToken(subject string) (*jwt.Token, error)
+	log.ContextLogger
```
Comment (bboozzoo) [inline jwt.go]: Looks like we can drop `log.ContextLogger` here. It does not seem to be used at all.

**Comment by mchalski** [comment]
@bboozzoo fixed and rebased - I'm merging this.

**Pull request merged**